config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java

```diff
@@ -0,0 +1,95 @@
+package com.typesafe.config.impl;
+
+import com.typesafe.config.ConfigNode;
+
+import java.util.*;
+
+final class ConfigNodeComplexValue implements ConfigNode, ConfigNodeValue {
+    private ArrayList<ConfigNode> children;
+    final private LinkedHashMap<Path, Integer> map = new LinkedHashMap<>();
+    final ArrayList<Integer> keyValueIndexes;
+
+    ConfigNodeComplexValue(Collection<ConfigNode> children) {
+        this.children = new ArrayList(children);
+        keyValueIndexes = new ArrayList<Integer>();
+
+        // Construct the list of indexes of Key-Value nodes. Do this
+        // in reverse order, since all but the final duplicate will be removed.
+        for (int i = this.children.size() - 1; i >= 0; i--) {
+            ConfigNode currNode = this.children.get(i);
+            if (currNode instanceof ConfigNodeKeyValue) {
+                keyValueIndexes.add(i);
+            }
+        }
+    }
+
+    public ArrayList<ConfigNode> children() {
+        return children;
+    }
+
+    public String render() {
+        StringBuilder renderedText = new StringBuilder();
+        for (ConfigNode child : children) {
+            renderedText.append(child.render());
+        }
+        return renderedText.toString();
```

Comment:
@havocp I really like this idea, and have an implementation in place locally. However, I've run into an issue.

As the code currently stands in this PR, only `ConfigNodeBasic`, `ConfigNodeKey,`, and `ConfigNodeSimpleValue` subclass `AbstractConfigNode`, as it's built to be a single-token node. ALL nodes implement the `ConfigNode` interface, and `ConfigNodeSimpleValue` and `ConfigNodeComplexValue` both implement the `ConfigNodeValue` interface, which extends the `ConfigNode` interface. To get this implementation, I definitely think it's best to make all node types subclass `AbstractConfigNode`, which I've done, since the tokens are an internal implementation detail and so the `tokens()` method should not be public-facing. 

However, this is leading to some casting issues. I've changed all nodes with children to keep an `ArrayList` of `AbstractConfigNode`s since they're going to need to produce the tokens of their children using the `tokens` method. However, I still need some way to differentiate the value node types from the other node types. `ConfigNodeValue` can't extend `AbstractConfigNode` since it's an interface, and thus, a `ConfigNodeValue` cannot be cast to an `AbstractConfigNode`, which means they can't be added to the list of children on a node. I could keep the list of children as a list of `ConfigNode`s instead of `AbstractConfigNode`s and have `ConfigNodeValue` extend `ConfigNode`, but then every single node is going to need to be casted when the `tokens()` function is called.

My solution to this is to have an abstract class called `AbstractConfigNodeValue`. This class does nothing and has no special variables, but subclasses `AbstractConfigNode`, which allows type differentiation. However, I'm feeling pretty iffy on this, since it requires the existence of a class that doesn't do anything except allow for type differentiation. I run into the same issue with the `ConfigNodeValue` interface, since that extends `ConfigNode` and doesn't do anything differently from `ConfigNode`. 

@havocp or @cprice404 do either of you have any suggestions on how to get around this problem? It seems like we need to differentiate the value nodes for parameter passing when setting values in `ConfigNodeKeyValue` and `ConfigNodeComplexValue`, but I can't think of another way to differentiate them apart from this, especially since `ConfigNodeComplexValue` and `ConfigNodeSimpleValue` have almost nothing in common apart from the fact that they are both nodes.